x/oracle/keeper/historic_price.go

```diff
@@ -0,0 +1,201 @@
+			Add(prices[lenPrices/2].ExchangeRates.ExchangeRate).
+			QuoInt64(2)
+	} else {
+		medianPrice = prices[lenPrices/2].ExchangeRates.ExchangeRate
+	}
+
+	return medianPrice
+}
+
+// GetMedian returns the median price of a given denom.
+func (k Keeper) GetMedian(
+	ctx sdk.Context,
+	denom string,
+) (sdk.Dec, error) {
+	store := ctx.KVStore(k.storeKey)
+	denom = strings.ToUpper(denom)
+	bz := store.Get(types.GetMedianKey(denom))
+	if bz == nil {
+		return sdk.ZeroDec(), sdkerrors.Wrap(types.ErrUnknownDenom, denom)
+	}
+
+	decProto := sdk.DecProto{}
+	k.cdc.MustUnmarshal(bz, &decProto)
+
+	return decProto.Dec, nil
+}
+
+// setMedian uses all the historic prices of given denom to set the
+// the median price of that denom in the store.
+func (k Keeper) setMedian(
+	ctx sdk.Context,
+	denom string,
+) {
+	store := ctx.KVStore(k.storeKey)
+
+	historicPrices := k.GetHistoricPrices(ctx, denom)
+
+	median := median(historicPrices)
+
+	bz := k.cdc.MustMarshal(&sdk.DecProto{Dec: median})
+	denom = strings.ToUpper(denom)
+	store.Set(types.GetMedianKey(denom), bz)
+}
+
+// GetHistoricPrice returns the historic price of a denom at a given
+// block.
+func (k Keeper) GetHistoricPrice(
+	ctx sdk.Context,
+	denom string,
+	blockNum uint64,
+) (types.HistoricPrice, error) {
+	store := ctx.KVStore(k.storeKey)
+	denom = strings.ToUpper(denom)
+	bz := store.Get(types.GetHistoricPriceKey(denom, blockNum))
+	if bz == nil {
+		return types.HistoricPrice{}, sdkerrors.Wrap(types.ErrUnknownDenom, denom)
+	}
+
+	var historicPrice types.HistoricPrice
+	k.cdc.MustUnmarshal(bz, &historicPrice)
+
+	return historicPrice, nil
+}
+
+// GetHistoricPrices returns all the historic prices of a given denom.
+func (k Keeper) GetHistoricPrices(
+	ctx sdk.Context,
+	denom string,
+) []types.HistoricPrice {
+	denom = strings.ToUpper(denom)
+	historicPrices := []types.HistoricPrice{}
+
+	k.IterateHistoricPrices(ctx, denom, func(exchangeRate sdk.Dec, blockNum uint64) bool {
+		historicPrices = append(historicPrices, types.HistoricPrice{
+			ExchangeRates: types.ExchangeRateTuple{
+				Denom:        denom,
+				ExchangeRate: exchangeRate,
+			},
+			BlockNum: blockNum,
+		})
+
+		return false
+	})
+
+	return historicPrices
+}
+
+// IterateHistoricPrices iterates over historic prices of a given
+// denom in the store.
+func (k Keeper) IterateHistoricPrices(
+	ctx sdk.Context,
+	denom string,
+	handler func(sdk.Dec, uint64) bool,
+) {
+	store := ctx.KVStore(k.storeKey)
+
+	iter := sdk.KVStorePrefixIterator(store, types.KeyPrefixHistoricPrice)
+	defer iter.Close()
+
+	for ; iter.Valid(); iter.Next() {
+		key := iter.Key()
+		historicPriceDenom := string(key[len(types.KeyPrefixExchangeRate) : len(key)-9])
+		if historicPriceDenom != denom {
+			continue
+		}
+		historicPrice := types.HistoricPrice{}
+
+		k.cdc.MustUnmarshal(iter.Value(), &historicPrice)
+		if handler(historicPrice.ExchangeRates.ExchangeRate, historicPrice.BlockNum) {
```

Comment:
This is a bit repetitive, in another PR do you think it would make sense to make the `HistoricPrice` struct look like this?

```golang
type HistoricPrice struct {
	Rate ExchangeRateTuple
	BlockNum      uint64
}
```

Reply:
I think it already is like that. Do you mean like this?
```golang
type HistoricPrice struct {
	Rate          sdk.Dec
	BlockNum      uint64
}
```

Reply:
Yeah, if the denom is a part of the key we don't need to read this